qa/rpc-tests/p2p-segwit.py

```diff
@@ -0,0 +1,1575 @@
+        temp_utxo.append(UTXO(tx2.sha256, 0, tx2.vout[0].nValue))
+
+        # Spend everything in temp_utxo back to an OP_TRUE output.
+        tx3 = CTransaction()
+        total_value = 0
+        for i in temp_utxo:
+            tx3.vin.append(CTxIn(COutPoint(i.sha256, i.n), b""))
+            tx3.wit.vtxinwit.append(CTxinWitness())
+            total_value += i.nValue
+        tx3.wit.vtxinwit[-1].scriptWitness.stack = [witness_program]
+        tx3.vout.append(CTxOut(total_value - 1000, CScript([OP_TRUE])))
+        tx3.rehash()
+        # Spending a higher version witness output is not allowed by policy,
+        # even with fRequireStandard=false.
+        self.test_node.test_transaction_acceptance(tx3, with_witness=True, accepted=False)
+        self.test_node.sync_with_ping()
+        with mininode_lock:
+            assert(b"reserved for soft-fork upgrades" in self.test_node.last_reject.reason)
+
+        # Building a block with the transaction must be valid, however.
+        block = self.build_next_block()
+        self.update_witness_block_with_transactions(block, [tx2, tx3])
+        self.test_node.test_witness_block(block, accepted=True)
+        sync_blocks(self.nodes)
+
+        # Add utxo to our list
+        self.utxo.append(UTXO(tx3.sha256, 0, tx3.vout[0].nValue))
+
+
+    def test_premature_coinbase_witness_spend(self):
+        print("\tTesting premature coinbase witness spend")
+        block = self.build_next_block()
+        # Change the output of the block to be a witness output.
+        witness_program = CScript([OP_TRUE])
+        witness_hash = sha256(witness_program)
+        scriptPubKey = CScript([OP_0, witness_hash])
+        block.vtx[0].vout[0].scriptPubKey = scriptPubKey
+        # This next line will rehash the coinbase and update the merkle
+        # root, and solve.
+        self.update_witness_block_with_transactions(block, [])
+        self.test_node.test_witness_block(block, accepted=True)
+
+        spend_tx = CTransaction()
+        spend_tx.vin = [CTxIn(COutPoint(block.vtx[0].sha256, 0), b"")]
+        spend_tx.vout = [CTxOut(block.vtx[0].vout[0].nValue, witness_program)]
+        spend_tx.wit.vtxinwit.append(CTxinWitness())
+        spend_tx.wit.vtxinwit[0].scriptWitness.stack = [ witness_program ]
+        spend_tx.rehash()
+
+        # Now test a premature spend.
+        self.nodes[0].generate(98)
+        sync_blocks(self.nodes)
+        block2 = self.build_next_block()
+        self.update_witness_block_with_transactions(block2, [spend_tx])
+        self.test_node.test_witness_block(block2, accepted=False)
+
+        # Advancing one more block should allow the spend.
+        self.nodes[0].generate(1)
+        block2 = self.build_next_block()
+        self.update_witness_block_with_transactions(block2, [spend_tx])
+        self.test_node.test_witness_block(block2, accepted=True)
+        sync_blocks(self.nodes)
+
+
+    def test_signature_version_1(self):
```

Comment:
For clarity, we should probably change this to SIGVERSION_WITNESS_V0 to match the enum that was added to the code, but that enum has value 1 (which is where "signature version 1" comes from).